test/README.md

```diff
@@ -180,26 +180,105 @@ quick changes, such as adding debugging output to JavaScript.
-machine. The nondestructive test should clean up after itself and restore the
-state of the machine, such that the next nondestructive test is not impacted.
-
-A fast running test suite is more important than independent,
-small test cases.
-
-Thus, it is OK for tests to be long.  Starting the test machine is so slow that
-we should run as many checks within a single session as make sense. Note that
-nondestructive tests do not suffer from this, and are much quicker.
-
-Still, within a long test, try to have independent sections, where
-each section returns the machine to more or less the state that it was
-in before the section.  This makes it easier to run these sections
-ad-hoc when doing incremental development.
+### Writing tests for bugs
+
+Start with an integration test which reproduces the issue, then develop
+against that to make the test pass.
+
+See for example d23a4647aed91d2ae3ba117919aab8172117a383
+
+### Writing tests for features
+
+For a new feature, our goal is to cover the new functionalities, our goal is
+usually to cover both the "positive" functionalities, as well as a reasonable
+set of corner cases and error conditions (guided by coverage reports)
+
+See for example aa5f640fc34ec6294383a90623c97b47df2fad0e
+
+### Writing a test
+
+The integration tests are located in the `verify` directory and usually named
+test-$page-$component for example check-system-terminal. All tests are written
+using Python's `unittest` library and inherit from `MachineCase`. Below
+is an example of a test which logs in and verifies that expected HTML classes
+are there:
+
+```python
+class TestLogin(MachineCase):
+	def testBasic(self):
+		b = self.browser
+		m = self.machine
+
+		b.open("/system")
+		b.wait_visible("#login")
+		b.set_val("#login-user-input", "admin")
+		b.set_val("#login-password-input", "foobar")
+		b.click("#login-button")
+		b.enter_page("/system")
+
+		b.wait_visible("#content")
+		b.wait_visible('#system_information_os_text')
+```
+
+We define a new test class `TestLogin` which inherits from the `MachineCase`
+class, this class does a few things for us. It gives us a `self.machine`
+variable which a `TestVM` and how we can interact with the test virtual
+machine. The `self.browser` variable is an instance of the `Browser` class
+which is how we interact with the test browser and control it.
+
+Cockpit tests on multiple distributions and versions, some features are
+specific to one particular distro for such scenario's Cockpit has multiple ways
+to skip a test.
+
+To only test on RHEL the `onlyImage` decorator can be used
+```
+@onlyImage('rhel-*')
+class TestRhelFeature(MachineCase):
+    ...
+```
+
+To skip a test because for example RHEL lacks a feature:
+
+```python
+skipImage("no btrfs support on RHEL", "rhel-*")`
+class TestBtrfs(MachineCase):
+    ...
+```
+
+Other commonly used decorators:
+
+* skipOstree - skip a test on an OSTree based distribution
+* skipMobile - skip a test on a mobile resolution
```

Comment:
write them as decorators wrapped by backticks:


* `@skipOstree` - skip a test on an OSTree based distribution
* `@skipMobile` - skip a test on a mobile resolution

Reply:
add links to source code where users can read the docstrings?